common/pci.c

```diff
@@ -0,0 +1,160 @@
+                if (0 == func)
+                    break;
+                continue;
+            }
+
+            new_dev = kzalloc(sizeof(*new_dev));
+            BUG_ON(!new_dev);
+
+            new_dev->segment = 0;
+            new_dev->bus = bus;
+            new_dev->dev = dev;
+            new_dev->func = func;
+            new_dev->vendor_id = cfg_val & 0xFFFF;
+            new_dev->device_id = cfg_val >> 16;
+
+            cfg_val = pci_cfg_read(bus, dev, func, 0x1);
+            new_dev->command = cfg_val & 0xFFFF;
+            new_dev->status = cfg_val >> 16;
+
+            cfg_val = pci_cfg_read(bus, dev, func, 0x2);
+            new_dev->subclass = (cfg_val & 0xFF0000) >> 16;
+            new_dev->class = cfg_val >> 24;
+
+            cfg_val = pci_cfg_read(bus, dev, func, 0x3);
+            new_dev->hdr_type = (cfg_val & 0xFF0000) >> 16;
+
+            if (new_dev->status & PCI_STATUS_CAP_LIST) {
+                new_dev->cap_ptr = pci_cfg_read8(bus, dev, func, 0xD);
+            }
+
+            new_dev->bridge = bridge;
+            snprintf(&new_dev->bdf_str[0], sizeof(new_dev->bdf_str), "%02x:%1x.%1x", bus,
+                     dev, func);
+
+            list_add_tail(&new_dev->list, &pci_list);
+
+            printk("pci: found device @ %s\n", &new_dev->bdf_str);
+
+            if (pci_dev_hdr_type(new_dev) == PCI_HDR_PCI_BRIDGE) {
+                uint8_t secondary, subordinate;
+
+                cfg_val = pci_cfg_read(bus, dev, func, 0x6);
+                secondary = (cfg_val & 0xFF00) >> 8;
+                subordinate = (cfg_val & 0xFF0000) >> 16;
+
+                for (uint8_t b = secondary; b <= subordinate; b++) {
+                    probe_pci_bus(b, 0, new_dev);
+                }
+            }
+
+            if (!pci_dev_is_multifunc(new_dev))
+                break;
+        }
+    }
+}
+
+static void probe_pci(void) {
+    pcidev_t *host_bridge;
+    uint32_t cfg_val;
+
+    if ((pci_cfg_read(0, 0, 0, 0) & 0xFFFF) == 0xFFFF) {
+        printk("pci: non-existent host bridge @ 00.0.0\n");
+        return;
+    }
+
+    if ((pci_cfg_read(0, 0, 0, 2) & 0xFFFF0000) != 0x06000000) {
+        printk("pci: expected host bridge class code @ 00.0.0\n");
+        return;
+    }
+
+    // Found the host bridge, initialize it
+    host_bridge = kzalloc(sizeof(*host_bridge));
+    BUG_ON(!host_bridge);
+
+    host_bridge->segment = 0;
+    host_bridge->vendor_id = pci_cfg_read(0, 0, 0, 0) & 0xFFFF;
+    host_bridge->device_id = pci_cfg_read(0, 0, 0, 0) >> 16;
+
+    cfg_val = pci_cfg_read(0, 0, 0, 1);
+    host_bridge->command = cfg_val & 0xFFFF;
+    host_bridge->status = cfg_val >> 16;
+
+    host_bridge->class = 0x06;
+    host_bridge->subclass = 0x00;
+    host_bridge->bridge = NULL;
+
+    snprintf(&host_bridge->bdf_str[0], sizeof(host_bridge->bdf_str), "00:0.0");
+
+    list_add_tail(&host_bridge->list, &pci_list);
+
+    printk("pci: found host bridge @ %s\n", &host_bridge->bdf_str);
```

Comment:
please move to `pci_init()`

Reply:
this would cause the printk to occur after all the other printks that indicate devices have been found. Is that alright?

Reply:
I would add a loop over the pci devs list in the `pci_init()` displaying everything that was found. Thereby, the main code will be printk-free. Also the loop can be easily guarded with opt_debug.

Reply:
Got it thanks, will update